File: src/main/java/kr/devport/api/domain/wiki/repository/WikiChatMessageRepository.java
package kr.devport.api.domain.wiki.repository;

import kr.devport.api.domain.wiki.entity.WikiChatMessage;
import kr.devport.api.domain.wiki.entity.WikiChatSession;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public interface WikiChatMessageRepository extends JpaRepository<WikiChatMessage, Long> {

    List<WikiChatMessage> findBySessionOrderByCreatedAtAsc(WikiChatSession session);

    List<WikiChatMessage> findTop20BySessionOrderByCreatedAtDesc(WikiChatSession session);

    int countBySession(WikiChatSession session);

    @Query("SELECT m.session.id, COUNT(m) FROM WikiChatMessage m WHERE m.session IN :sessions GROUP BY m.session.id")
    List<Object[]> countBySessionsGrouped(@Param("sessions") List<WikiChatSession> sessions);

    default Map<Long, Long> countMapForSessions(List<WikiChatSession> sessions) {
        return countBySessionsGrouped(sessions).stream()
                .collect(Collectors.toMap(row -> (Long) row[0], row -> (Long) row[1]));
    }
}

File: wiki/repository-jpa/src/main/kotlin/kr/devport/api/domain/wiki/repository/WikiChatMessageRepository.kt
package kr.devport.api.domain.wiki.repository

import kr.devport.api.domain.wiki.entity.WikiChatMessage
import kr.devport.api.domain.wiki.entity.WikiChatSession
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param

interface WikiChatMessageRepository : JpaRepository<WikiChatMessage, Long> {
    fun findBySessionOrderByCreatedAtAsc(session: WikiChatSession): List<WikiChatMessage>

    fun findTop20BySessionOrderByCreatedAtDesc(session: WikiChatSession): List<WikiChatMessage>

    fun countBySession(session: WikiChatSession): Int

    @Query("SELECT m.session.id, COUNT(m) FROM WikiChatMessage m WHERE m.session IN :sessions GROUP BY m.session.id")
    fun countBySessionsGrouped(
        @Param("sessions") sessions: List<WikiChatSession>,
    ): List<Array<Any>>

    fun countMapForSessions(sessions: List<WikiChatSession>): Map<Long, Long> =
        countBySessionsGrouped(sessions).associate { row -> (row[0] as Long) to (row[1] as Long) }
}
